Migrate workgraph data from `base.extras` to `base.attributes`

superstar54 opened the pull request:
# Engine

## Inputs
In WorkGraph engine, the `workgraph data`, including the tasks, links, inputs and outputs of the tasks etc, are passed as inputs of the WorkGraphEngine.

This `workgraph data` will be used in:
- execution of the WorkGraph
- re-load the WorkGraph from the database

In the old design, the `workgraph data` is stored in the `base.extras` of the node. However, there are several drawbacks:
- If the `base.extas` is deleted, one can not restore the WorkGraph
- could have a security issue if the data is modifed by other people.

In this PR, we save the `workgraph data` as an attribute of the node (`WorkGraphNode`), when creating the process (`WorkGraphEngine`). This attribute will be immutable when the WorkGraph is finished.


## Updatable attributes
When the `WorkGraphNode` is created (in the `on_create` method), it is also stored, and its attributes become immutable. However, some of the data need updating during the execution of the WorkGraph, this includes the runtime information, e.g.,:

- state of the task
- action (e.g., pause, reset) of the task
- process of the task, because


To achieve this, we have to define the attributes that can be updated in the `_updatable_attributes` list, and save these runtime info as the attributes of the node.

## Issues
For the moment, the default logger of the `WorkGraphNode` has a `WARNING` level, however, we need `REPORT` level. I hardcoded the `logger` for the `WorkGraphNode` using the same one as in the `WorkChainNode` to fix this. We need to fix this in the future.

superstar54 committed:
Use WorkGraphNode to store inputs of WorkGraph and the state of the task.

- Update the WorkGraphNode with `workgraph_data`, `task_states`, `task_processes`, `task_actions` keys
- Use `metadata.workgraph_data` to hold the `wgdata`
- Get/set the state of the task using the attributes

superstar54 committed:
use `workgraph_data` input directly

superstar54 committed:
Use _raw_inputs to create the `workgraph_data` attribute

also fix the logger of the WorkGraphNode

superstar54 committed:
Fix tests

superstar54 committed:
Fix restart

superstar54 committed:
rename `set_task_state_info` to `set_task_runtime_info`

GeigerJ2 reviewed (commented):
Not familiar enough with the whole code base to give a founded review, I think. Changes look good to me. I saw @agoscinski requested changes, so I'll only comment here for now.

superstar54 committed:
Apply review suggestions:

- format the `engine` using the rst file
- fix the type annotation
- add setter in the orm.WorkGraph

agoscinski reviewed (approved):
Thanks!

GeigerJ2 reviewed (approved):
LGTM
